037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,177 @@
+This new CRD should be quite similar to the `KafkaRebalance`.
+The cluster operator has to copy over the configuration from the template to the actual rebalancing resource.
+Compared to the rejected alternative, it’s about exporting the rebalancing configuration out of the Kafka custom resource avoiding to pollute that resource and increase the size.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalanceTemplate
+metadata:
+  name: my-rebalance-template
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  # NOTE: mode and brokers fields are not available 
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+The user has the flexibility to specify goals as well as all the other options like he already does today on a `KafkaRebalance` custom resource.
+The cluster operator will use that configuration for the `KafkaRebalance` but adding the `spec.mode` and `spec.brokers` accordingly.
+
+It would also contain a `spec.autoApproval` field, as described by this [proposal](https://github.com/strimzi/proposals/pull/56), to specify if auto-approval is desired or not when auto-rebalancing is enabled.
+The template to be used is referenced by a new `spec.kafka.autoRebalance.template` field and the cluster operator uses that template to create a corresponding `KafkaRebalance` custom resource.
+If the reference is missing, the cluster operator can create an “empty” `KafkaRebalance` custom resource (which means the optimization proposal will be based on Cruise Control default configuration). It will behave as the rejected alternative.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      mode: [running, scaling] # or just one of the two
+      template: # reference to a KafkaRebalanceTemplate custom resource
+```
+
+### Flow
+
+Auto-rebalancing on scaling up.
+
+1. The user scales up the cluster by increasing the `spec.kafka.replicas` field.
+2. The cluster operator takes care of it by scaling up the cluster as usual, by adding additional brokers.
+3. If `spec.kafka.autoRebalance.mode` contains the `scaling` value, the cluster operator creates a `KafkaRebalance` custom resource.
+   * It will specify `spec.mode` as `add-brokers` and `spec.brokers` with the list of brokers already added to the cluster.
+4. The cluster operator asks Cruise Control to generate an optimization proposal via the REST API.
+5. When the optimization proposal is ready, the cluster operator checks if the user has specified auto-approval or not:
```

Comment:
The `KafkaRebalanceAssemblyOperator` has its own internal polling (to the CC REST API) for checking that a proposal is ready. It doesn't do that on each reconcile because it would mean to know that a proposal is ready after looooong time (by default 2 minutes).